beacon_node/beacon_chain/src/shuffling_cache.rs

```diff
@@ -45,18 +51,31 @@ impl CacheItem {
     }
 }
 
-/// Provides an LRU cache for `CommitteeCache`.
+/// Provides a cache for `CommitteeCache`.
 ///
 /// It has been named `ShufflingCache` because `CommitteeCacheCache` is a bit weird and looks like
 /// a find/replace error.
-pub struct ShufflingCache {
-    cache: LruCache<AttestationShufflingId, CacheItem>,
+pub struct ShufflingCache<T: BeaconChainTypes> {
+    cache: HashMap<AttestationShufflingId, CacheItem>,
+    cache_size: usize,
+    head_shuffling_decision_root: Hash256,
+    slot_clock: T::SlotClock,
+    logger: Logger,
 }
 
-impl ShufflingCache {
-    pub fn new(cache_size: usize) -> Self {
+impl<T: BeaconChainTypes> ShufflingCache<T> {
+    pub fn new(
+        cache_size: usize,
+        head_shuffling_decision_root: Hash256,
```

Comment:
I wonder if rather than a single `head_shuffling_decision_root` which only applies to the current epoch, whether we should _also_ store the head's decision root for the next epoch. In smooth conditions the head's next epoch shuffling will be stable, because it is determined by the last block of epoch `N - 1` (where the head is in epoch `N`).

To achieve this we could store a `BlockShufflingIds` for the head, which contains two `AttesterShufflingId`s. I think this would also allow us to remove the internal slot clock, as the head's `BlockShufflingIds` contain the relevant epochs.

Reply:
Jimmy and I had a chat about this and he's planning to store:

- Current epoch shuffling root
- Next epoch shuffling root
- Head root (for all following shuffling roots)

Reply:
Note that the VC will request next epoch duties regularly, so the head's next epoch shuffling will normally be in the cache. The function that puts it there is:

https://github.com/sigp/lighthouse/blob/693886b94176faa4cb450f024696cb69cda2fe58/beacon_node/beacon_chain/src/beacon_chain.rs#L1412-L1417

Reply:
@paulhauner Perfect! Thought you might be onto it already

Reply:
I'm not sure if we could remove the internal slot clock. I understand that the "head shuffling" always has a shuffling key with an epoch that is based on the current wall-clock slot. I don't know how else to get that, other than with a slot clock.

Do we have a different understanding of the "head shuffling" @michaelsproul?

Reply:
I was thinking if the `BlockShufflingIds` struct is built outside of the cache (in `canonical_head.rs`) then passed to `update_head_shuffling_decision_root`, then the cache doesn't need to re-derive the time. It can just compare the shuffling IDs directly, can't it?

Reply:
The scenario I'm thinking is:

1. The head is in epoch N and `update_head_shuffling_decision_root` is called and provides `(current_shuffling_root, next_shuffling_root, head_root)`
1. No more blocks are set as the head.
1. In epoch N we should retain the shufflings for `{ epoch: N, root: current_shuffling_root }`
1. In epoch N+1 we should retain the shuffling for `{ epoch: N+1, root: next_shuffling_root }`
1. In epoch N+2 we should retain the shuffling for `{ epoch: N+2, root: head_root }`

And we need to have a slot clock so we can if we're in scenarios (3), (4) or (5).

Reply:
`update_head_shuffling_decision_root` is called in `after_new_head`, so if we don't get any new head updates, i think shuffling cache won't know what epoch its in without the `slot_clock`.

Reply:
Now I'm thinking about this we may want to retain the shuffling for the current epoch *and* the previous epoch, since we're likely to be processing attestations which refer to either of those.

Reply:
Yeah was just thinking about the same thing. Also if we just retain shuffling for all three of them, then we won't need `slot_clock`?

Reply:
Oh yeah, good point. I think it's still possible, but maybe less clear. The shuffling IDs eligible for pruning would be:

```rust
.filter(|shuffling_id| {
  Some(shuffling_id) != self.head_shuffling_ids.id_for_epoch(shuffling_id.shuffling_epoch)
})
```

In case (1) if `{ epoch: N, root: current_shuffling_root }` is in the cache, then it will be equal to the shuffling ID derived from the head for epoch `N` and be ineligible for pruning. Similarly for case (2).

In case (3) if there are any cache entries for any epoch >= N + 2 with the head as the decision block root they will be retained. This is slightly different from what we'd get with the explicit clock, in that it would also protect the shuffling at epoch N + 3 during epoch N + 2. The disadvantage is that the cache _could_ get flooded with future entries based on the head, although I think the BN and VC should already guard against this happening (i.e. we wouldn't put entries in the cache for epochs beyond the next epoch).

I don't mind which way we go, the `BlockShufflingIds` approach is maybe nice for code re-use, but arguably the clock approach is clearer

Reply:
I now understand Michael's solution. It's elegant and I prefer it to my proposal.

As discussed offline, adding a `previous: Option<AttestationShufflingId>` field to `BlockShufflingIds` and using Michael's `filter` example (see prev comment) is now my preference.

Reply:
Thanks for the review and suggestions ☺️ 
I've pushed an update based on Michael's solution and updated the tests as well to be a bit more realistic.